witness/golang/cmd/witness/internal/witness/witness_test.go

```diff
@@ -0,0 +1,287 @@
+	}
+	sigVs := []note.Verifier{logV}
+	log := LogInfo{
+		SigVs: sigVs,
+		LogV:  logverifier.New(h),
+	}
+	m := map[string]LogInfo{logID: log}
+	opts := Opts{
+		DB:        d,
+		Signer:    ns,
+		KnownLogs: m,
+	}
+	return opts, nil
+}
+
+func newOptsAndKeys(d *sql.DB, logID string, logPK string) (Opts, error) {
+	wSK, _, err := note.GenerateKey(rand.Reader, "witness")
+	if err != nil {
+		return Opts{}, fmt.Errorf("couldn't generate witness keys")
+	}
+	return newOpts(d, logID, logPK, wSK)
+}
+
+func newWitness(t *testing.T, d *sql.DB, logID string, logPK string) *Witness {
+	opts, err := newOptsAndKeys(d, logID, logPK)
+	if err != nil {
+		t.Fatalf("couldn't create witness opt struct: %v", err)
+	}
+	w, err := New(opts)
+	if err != nil {
+		t.Fatalf("couldn't create witness: %v", err)
+	}
+	return w
+}
+
+// dh is taken from https://github.com/google/trillian/blob/master/merkle/logverifier/log_verifier_test.go.
+func dh(h string, expLen int) []byte {
+	r, err := hex.DecodeString(h)
+	if err != nil {
+		panic(err)
+	}
+	if got := len(r); got != expLen {
+		panic(fmt.Sprintf("decode %q: len=%d, want %d", h, got, expLen))
+	}
+	return r
+}
+
+func mustCreateDB(t *testing.T) (*sql.DB, func() error) {
+	t.Helper()
+	db, err := sql.Open("sqlite3", ":memory:")
+	if err != nil {
+		t.Fatalf("failed to open temporary in-memory DB: %v", err)
+	}
+	return db, db.Close
+}
+
+func TestGetChkpt(t *testing.T) {
+	for _, test := range []struct {
+		desc    string
+		setID   string
+		queryID string
+		c       Chkpt
+		want    string
+	}{
+		{
+			desc:    "happy path",
+			setID:   "testlog",
+			queryID: "testlog",
+			c:       initChkpt,
+			want:    "there",
+		}, {
+			desc:    "other log",
+			setID:   "testlog",
+			queryID: "otherlog",
+			c:       initChkpt,
+			want:    "not there",
+		}, {
+			desc:    "nothing there",
+			setID:   "testlog",
+			queryID: "testlog",
+			c:       initChkpt,
+			want:    "not there",
+		},
+	} {
+		t.Run(test.desc, func(t *testing.T) {
+			d, closeFn := mustCreateDB(t)
+			defer closeFn()
+			ctx := context.Background()
+			// Set up log keys and sign checkpoint.
+			logSK, logPK, err := note.GenerateKey(rand.Reader, test.setID)
+			if err != nil {
+				t.Errorf("couldn't generate log keys: %v", err)
+			}
+			signed, err := signChkpts(logSK, []string{string(test.c.Raw)})
+			if err != nil {
+				t.Errorf("couldn't sign checkpoint: %v", err)
+			}
+			test.c.Raw = signed[0]
+			// Set up witness keys and other parameters.
+			wSK, wPK, err := note.GenerateKey(rand.Reader, "witness")
+			if err != nil {
+				t.Errorf("couldn't generate witness keys: %v", err)
+			}
+			opts, err := newOpts(d, test.setID, logPK, wSK)
+			if err != nil {
+				t.Errorf("couldn't create witness opts: %v", err)
+			}
+			w, err := New(opts)
+			if err != nil {
+				t.Fatalf("couldn't create witness: %v", err)
+			}
+			// Set a checkpoint for the log if we want to for this test.
+			if test.desc != "nothing there" {
```

Comment:
A less brittle approach would be to have another member of the test struct which gates this behaviour more directly - e.g. it could be `logCP *api.Checkpoint` which is left unset if you don't want this `Update` to be called

Reply:
Cool, done.